### PR TITLE
Consolidate message reply buttons

### PR DESCRIPTION
**Pull request by StCyr**
This PR intends to replace the "Send" and "Forward" buttons by a single action menu ("..." menu) providing these functionalities plus the "Reply all" function

To do

* [x] Position actions at the top right
* [x] Expand the popover menu to a vertical group of buttons

### CODE REVIEW

**Comment by StCyr** [comment]
This is now looking good.

Except for the icons and positioning of the "reply", "reply all", and "forward" action buttons.

TODO:

1. Add reply, reply all, and forward icons and make use of them
   => Don't really know how to do this
2. Make these action buttons display in a row rather than in a column

File: src/components/Message.vue [hunk 1]
```diff
@@ -191,6 +188,32 @@ export default {
...
+					accountId: this.$route.params.accountId,
+					folderId: this.$route.params.folderId,
+					messageUid: 'reply',
+				},
+				query: {
+					uid: this.message.uid,
+				},
+			})
+		},
+		replyAllMessage() {
```
Comment (ChristophWurst) [inline src/components/Message.vue]: you could rename the method and add the `messageUid` as parameter :wink:

Comment (StCyr) [reply src/components/Message.vue]: Do you mean doing something like this?

```
<ActionButton icon="icon-mail" @click="reactOnMessage('reply')">{{ t('mail', 'Reply') }}</ActionButton>
<ActionButton icon="icon-delete" @click="reactOnMessage('replyAll')">{{ t('mail', 'Reply all') }}</ActionButton>
<ActionButton icon="icon-delete" @click="reactOnMessage('forward')">{{ t('mail', 'Forward') }}</ActionButton>
...
 reactOnMessage(action) {
                        this.$router.push({
                                name: 'message',
                                params: {
                                        accountId: this.$route.params.accountId,
                                        folderId: this.$route.params.folderId,
                                        messageUid: action,
                                },
                                query: {
                                        uid: this.message.uid,
                                },
                        })
}
```

I've tried but can't get it to work for some reason; As I said already, I'm a js noob :-).

Comment (ChristophWurst) [reply src/components/Message.vue]: Okay, strange. I expected that to just work :man_shrugging:

**Comment by StCyr** [comment]
- **Make these action buttons display in a row rather than in a column**

It doesn't look like ActionButton allows this. So, I'll probably have to revert to some `popovermenu` or something

**Comment by ChristophWurst** [comment]
> It doesn't look like ActionButton allows this. So, I'll probably have to revert to some `popovermenu` or something

cc @StCyr

**Comment by StCyr** [comment]
?

hello :-)

**Comment by StCyr** [comment]
Looks good to me like this

Additional note: I'm wondering if I must add some draft auto-saving function or if it's automagic

**Comment by ChristophWurst** [comment]
> hello :-)

sorry. I was @mentioning @skjnldsv but the github autocompletion picket the wrong user ;-)

**Comment by skjnldsv** [comment]
We thought about this with jancborchardt.
Let me open an issue on the repo :)

**Comment by ChristophWurst** [comment]
> I find my changes satisfactory now.

Every time I have that feeling @skjnldsv appears and rejects my pull requests :see_no_evil: 

Thanks for your patch. I will review it soonish :)

**Comment by ChristophWurst** [comment]
Updated the branch to our latest changes and resolved the conflicts :)

**Comment by ChristophWurst** [comment]
Updated the PR description with the open todos :)

**Comment by StCyr** [comment]
@ChristophWurst 

>Position actions at the top right

It's already the case

> Expand the popover menu to a vertical group of buttons

What do you mean by "vertical group of button"?

Do you mean something like this (imagine it with a proper layout)?

image

Can you or @nextcloud/design create a "reply all" and "forward" icon?

Best regards,

**Comment by ChristophWurst** [comment]
> What do you mean by "vertical group of button"?

of course I meant horizontal :speak_no_evil: , like this https://getbootstrap.com/docs/4.3/components/button-group/#basic-example



> Can you or @nextcloud/design create a "reply all" and "forward" icon?

Will try to get @jancborchardt's help on this

**Comment by StCyr** [comment]
unwrapping these 3 buttons outside a regular « ... » pop up menu button will probably make this app unusable on a regular smartphone though

**Comment by StCyr** [comment]
@ChristophWurst @skjnldsv @nextcloud/designers 

could someone have a look at my commit https://github.com/nextcloud/mail/pull/1774/commits/db0a6a3a7b47a8dfcd91a8d5a8d4a8328e9b081b ?

I'm trying to add 2 new icons to the app but can't figure out how to use them :-(

Cheers,

Cyrille

**Comment by StCyr** [comment]
> 
> 
> @ChristophWurst @skjnldsv @nextcloud/designers
> 
> could someone have a look at my commit [db0a6a3](https://github.com/nextcloud/mail/commit/db0a6a3a7b47a8dfcd91a8d5a8d4a8328e9b081b) ?
> 
> I'm trying to add 2 new icons to the app but can't figure out how to use them :-(
> 
> Cheers,
> 
> Cyrille

Nevermind, I had forgot to copy the css over to my nextcloud server, It couldn't have work :-p

**Comment by skjnldsv** [comment]
@StCyr use the `icon-black-white` mixin, it's easier and create compatibility with the dark theme automatically
```scss
@mixin icon-black-white('reply', 'mail', 1)
@mixin icon-black-white('reply-all', 'mail', 1)
@mixin icon-black-white('forward', 'mail', 1)
```

**Comment by StCyr** [comment]
> 
> 
> @StCyr use the `icon-black-white` mixin, it's easier and create compatibility with the dark theme automatically
> 
> ```css-scss
> @mixin icon-black-white('reply', 'mail', 1)
> @mixin icon-black-white('reply-all', 'mail', 1)
> @mixin icon-black-white('forward', 'mail', 1)
> ```

ok

**Comment by StCyr** [comment]
I think it's good now @ChristophWurst @jancborchardt @skjnldsv

**Comment by jancborchardt** [comment]
@StCyr nice work! Some feedback:
- The main "Reply" button should have the `class="primary"` so it appears blue and calls attention.
- The position of the elements seem off, they should be in the header and  vertically aligned with the subject and sender info
- The icons don’t show for me, do I need to integrate the vue-components for that or so?
- To save space for the title on mobile, the Reply button can be reduced to only the icon
- Not sure what’s happening with the rest of the layout? (Might be unrelated, but first time I’m seeing that)

mail screenshot

**Comment by StCyr** [comment]
worth reviewing please @jancborchardt :-)

**Comment by jancborchardt** [comment]
@StCyr cool, here’s the feedback :)

This is how it looks like for me now, let me know if there’s anything I’m missing:
mail

- [ ] The "Reply" button should have text in it too when the space permits like here on desktop.
- [ ] The icon and text inside the primary reply button should be white
- [ ] Maybe it’s better to put the buttons on the right because at the moment their position changes all the time?
- [ ] Currently there’s the case of the "Forward" icon being just there, which is difficult on touch interfaces as there’s no title. I’d say we should add the "Mark unread" and "Delete" in that menu too so there will always be a menu?
- [ ] When you hover the 3-dot-menu icon, there’s a strange double mark effect
  mail strange double hover

File: img/forward.svg [hunk 2]
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (skjnldsv) [inline img/forward.svg]: Please clean things up :)
You can use https://jakearchibald.github.io/svgomg/

Or our official scour `scour --create-groups --enable-id-stripping --enable-comment-stripping --shorten-ids --remove-metadata --strip-xml-prolog --no-line-breaks img/forward.svg img/forward.new.svg`

**Comment by StCyr** [comment]
Current status is:

- [ ] The "Reply" button should have text in it too when the space permits like here on desktop.
- [ ] Currently there’s the case of the "Forward" icon being just there, which is difficult on touch interfaces as there’s no title. I’d say we should add the "Mark unread" and "Delete" in that menu too so there will always be a menu?
  => This is not that trivial as we are in a message and "Mark unread" and "Delete" work with envelopes
- [ ] The 'reply.svg' icon that I've re-used has an opacity of 0,5 while the 'reply-all' and 'forward' icons that I've created have no opacity defined in their .svg definition. 
  => What should I do? (1) Set opacity=0,5 to the newly created icons? (2) remove opacity=0,5 from icon "reply.svg" (this will change how the icon is already used), (3) create a copy of "reply.svg" with opacity=1, (4) other?

**Comment by jancborchardt** [comment]
@StCyr it’s fine to convert the icon to full opacity. We should move to full opacity icons anyway.

**Comment by StCyr** [comment]
yeah! review please @jancborchardt

**Comment by jancborchardt** [comment]
@StCyr wow, reeeeeally awesome work! :) Everything looks super fine, only one little thing:
When the title or list of recipients is too long, you need to scroll to the right to see the buttons. Instead the title and list of recipients should get ellipsized to not necessitate a horizontal scrollbar and make the buttons visible on the top right:
mail app screen

**Comment by StCyr** [comment]
> 
> 
> @StCyr wow, reeeeeally awesome work! :) Everything looks super fine, only one little thing:
> When the title or list of recipients is too long, you need to scroll to the right to see the buttons. Instead the title and list of recipients should get ellipsized to not necessitate a horizontal scrollbar and make the buttons visible on the top right:
> mail app screen

yeah, the text's ellipsis doesn't work... But I think it didn't work before neither.

**Comment by jancborchardt** [comment]
@StCyr to fix the ellipsis, `#mail-message-header-fields h2, #mail-message-header-fields p` needs an additional: `max-width: calc(100vw - 144px);`

I can’t add it to the pull request as it is in your fork. You want to add it? :)

Also, now that you are member of the organization, you can also work in branches of this main repository.

**Comment by StCyr** [comment]
Of course I’ll add it :)

(I had try « calc(80%) » without success in FF’s responsive mode)

Also, I don’t see how my workflow will change when I work on this repository’s branches... except if I can write to master ofc.

**Comment by ChristophWurst** [comment]
> I can’t add it to the pull request as it is in your fork. You want to add it? :)

You can still push, even if it's a fork. Unless the person opening the branch explicitly didn't allow that. Just add the fork as new repo and push there.

**Comment by skjnldsv** [comment]
> (I had try « calc(80%) » without success in FF’s responsive mode)

Hum, best fixes should not have specific calc in them.
Please fix the layout instead so that the title is taking the max width available with flex. calc (100% - xxx) is always to avoid

**Comment by StCyr** [comment]
But ellipsis works only when width is not set as a percentage :(

**Comment by skjnldsv** [comment]
> But ellipsis works only when width is not set as a percentage :(

Yes, it does if the parent have the proper width. Then you can just use a max-width 100%. And let the other two properly push down the width to fit the page.

**Comment by StCyr** [comment]
implemented @jancborchardt's fix for ellipsing message title and addresses.

Remaining questions:

1. Why 144px in `max-width: calc(100vw - 144px);`? I've try "(100vw - 132px)" and it's also as nice;
2. To hide the reply button's label, I've used the following media query `@media only screen and (max-width: 600px)`. Is this value of 600px ok for you?;
3. Is there any other way than using `calc()`for ellipsing (@skjnldsv doesn't seem to like it)?

**Comment by skjnldsv** [comment]
> 3\. (@skjnldsv doesn't seem to like it)

Eheh, sorry! It's just harder to maintain and not proper css coding habits :)
Could I take a look if you allow me? :bowing_woman:

**Comment by StCyr** [comment]
> > 3. (@skjnldsv doesn't seem to like it)
> 
> Eheh, sorry! It's just harder to maintain and not proper css coding habits :)
> Could I take a look if you allow me? 🙇‍♀

Sure, go ahead.

Also, I'm on #nextcloud-design if you want to discuss

**Comment by skjnldsv** [comment]
Done :)

**Comment by jancborchardt** [comment]
Hmmm it doesn’t seem to work, and this shadow wasn’t there before?
mail header

**Comment by skjnldsv** [comment]
Can't reproduce
Capture d’écran_2019-09-04_20-02-51

**Comment by skjnldsv** [comment]
> Hmmm it doesn’t seem to work, and this shadow wasn’t there before?
> mail header

Ah! Found the culprit: https://github.com/nextcloud/mail/blob/7586fea141cf5636dd51d085a2a9c9a37061c443/css/mobile.css#L21

**Comment by StCyr** [comment]
hmmm, your fix doesn't look nice on FF/iOS @skjnldsv :

FF-iOS

**Comment by skjnldsv** [comment]
@StCyr please clear your cache :)
You still have the shadow but it's removed from this pr

**Comment by StCyr** [comment]
@skjnldsv 

I've tried very hard to clear all my caches but without succes.

Do you have it working properly on Firefox/iOS?

Cyr

**Comment by skjnldsv** [comment]
> Do you have it working properly on Firefox/iOS?

firefox OR iOS? Or Firefox ON iOS?

**Comment by StCyr** [comment]
> 
> 
> > Do you have it working properly on Firefox/iOS?
> 
> firefox OR iOS? Or Firefox ON iOS?

Firefox on iOS

**Comment by skjnldsv** [comment]
Weridly enough, this is indeed not working on iOS...

**Comment by StCyr** [comment]
So, how should we proceed to get this PR merged?

[ ] Accept @skjnldsv fix, too bad for iOs
[ ] Accept @jancborchardt fix, too bad for the css best practices
[ ] other

I also would like to remember you that I've:

1. Arbitrarly chosen a value of 600px in `@media only screen and (max-width: 600px)`
2. Changed the opacity of an existing icon from 0.5 to 1.0

This PR is expected to close a frequently reported problem (cfr: #1772, #1513, #192, and #287)

**Comment by skjnldsv** [comment]
> * [ ]  other

Fix skjnldsv typo? :D
There is a console error because I removed a line I shouldn't have https://github.com/nextcloud/mail/pull/1774/files#diff-6b24ffbce9ff03fc22f6bc8026cd750cL70

Also, somehow `box-sizing: content-box !important;` is not working on ios.
Fixing this

**Comment by skjnldsv** [comment]
So that was quite hard! :)
I rebased everything!

I also fixed the mobile display. Should be fine now (tested on safari ios)
Though @ChristophWurst the weird hack for responsive iframe still behave weirdly on narrow screens :thinking:

**Comment by ChristophWurst** [comment]
> Though @ChristophWurst the weird hack for responsive iframe still behave weirdly on narrow screens thinking

Something to fix at the Mail hackweek? ;)

**Comment by StCyr** [comment]
> I also fixed the mobile display. Should be fine now (tested on safari ios)

I confirm it works on FF/iOS. Great job @skjnldsv 👍

**Comment by skjnldsv** [comment]
> I confirm it works on FF/iOS. Great job @skjnldsv

All the merit to you :)
I just did some small css fixes :bowing_woman:

**Comment by StCyr** [comment]
I'm quite happy with this PR's result. Could you consider it for merging please?

Though, we may want to move the "Send" button (and attachment buttons also) at the top-right of the Composer component (as with the "Reply" and "Forward" buttons) as clicking on them now requires scrolling down to the bottom of the message, which may be quite long in case of a reply to a long email thread.

Cyrille

**Comment by skjnldsv** [comment]
@stCyr please squash your commits together first :)

**Comment by StCyr** [comment]
Is it ok like this @skjnldsv ?

**Comment by jancborchardt** [comment]
Good to go @ChristophWurst? :)

File: src/ReplyBuilder.js [hunk 3]
```diff
@@ -111,7 +111,7 @@ export const buildReplySubject = original => {
 // TODO: https://en.wikipedia.org/wiki/List_of_email_subject_abbreviations#Abbreviations_in_other_languages
 const forwardPrepends = ['fwd']
 
-export const buildFowardSubject = original => {
+export const buildForwardSubject = original => {
```
Comment (ChristophWurst) [inline src/ReplyBuilder.js]: thanks :see_no_evil:

File: src/components/Message.vue [hunk 4]
```diff
@@ -85,6 +102,7 @@ export default {
 			replyRecipient: {},
 			replySubject: '',
 			replyBody: '',
+			envelope: '',
```
Comment (ChristophWurst) [inline src/components/Message.vue]:
```suggestion
			envelope: undefined,
```

since an object?

File: src/components/Message.vue [hunk 5]
```diff
@@ -206,22 +254,93 @@ export default {
 				},
 			})
 		},
+		onToggleSeen() {
+			this.$store.dispatch('toggleEnvelopeSeen', this.envelope)
+		},
+		onDelete(e) {
+			// Don't try to navigate to the deleted message
+			e.preventDefault()
```
Comment (ChristophWurst) [inline src/components/Message.vue]: copy-paste error? I don't get why this is needed here. the action is just a button. it won't navigate by default

Comment (ChristophWurst) [reply src/components/Message.vue]: also this code below is unrelated, no?

Comment (StCyr) [reply src/components/Message.vue]: you're right the `e.preventDefault` isn't necessary here.

Though, I think the code below is needed: If the user deletes the currently displayed message we need to show the next/previous one.

**Comment by ChristophWurst** [comment]
Can't quite use Github at the moment as they have some [issues](https://www.githubstatus.com/), but will check later.

File: src/components/Message.vue (hunk 5 above)
Comment (ChristophWurst) [reply src/components/Message.vue]: Yep, but it's basically duplicated code …

merging now, but this has to be cleaned up later on

**Comment by deusoz** [comment]
This is great news.  I look forward to an updated Mail app so I can use these improvements.  I should see an updated Mail app available for NC16 right?

**Comment by ChristophWurst** [comment]
Hi,

not all recent releases are available for 16. Some require 17 [for technical reasons](https://github.com/nextcloud/mail/issues/2195#issuecomment-547021707).

**Pull request merged**